### PR TITLE
DDS-1097-trashbin_apib

### PR DESCRIPTION
**Pull request by dmlond**
Also generated public/apidocs/*.html

### CODE REVIEW

File: api_docs/trashbin.apib [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+## Trashbin Item instance [/trashbin/{object_id}/{object_kind}]
++ Parameters
+    + object_kind (string, required, `dds-file`) ... The kind of object.
+    + object_id (string, required, `777be35a-98e0-4c2e-9a17-7bc009f9b111`) ... The unique id of the object.
+
+### View Trashbin item [GET]
+>**Permission:** view_project
+- - - - -
+**Response messages:**
+* 200: Valid API Token in 'Authorization' Header
```
Comment (dmann) [inline api_docs/trashbin.apib]: Shouldn't this just be `200: Success`?

File: api_docs/trashbin.apib [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+                    "id": "ca29f7df-33ca-46dd-a015-92c46fdb6fd1",
+                    "name": "Knockout Mouse Project (KOMP)"
+                },
+                {
+                    "kind": "dds-folder",
+                    "id": "552be8c5-209d-4e3b-afaf-cb66686ffbyy",
+                    "name": "Sequencing Core"
+                }
+            ],
+            "is_deleted": false,
```
Comment (dmann) [inline api_docs/trashbin.apib]: Oh noes! A file in the trash bin that's not marked `is_deleted: true`!  😛

File: api_docs/trashbin.apib [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+- - - - -
++ If parent is not supplied in the payload, an attempt will be made to restore to the original parent.
++ An exception is returned if the parent is itself in the trash bin. You must restore the parent before you can restore a child to it.
++ `dds-file` and `dds-folder` objects can only be restored to a `dds-project` or `dds-folder` parent.
++ If a `dds-file` is restored, all deleted version history will be restored as well.
++ `dds-file-version` objects can only be restored to their original "owning" file.
++ If a `dds-file-version` is restored, all deleted version history of the original "owning" file will be restored as well.
+
+**Response messages:**
+* 201: Valid API Token in 'Authorization' Header
```
Comment (dmann) [inline api_docs/trashbin.apib]: Does this endpoint really return a `201`? Seems weird to have a `PUT` call return `201 Created`. If this is how it's been documented (and subsequently implemented) the entire time, don't bother changing it. Though, again, the `Valid API Token...` wording should go.

File: api_docs/trashbin.apib [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+## Purge Trashbin Item [/trashbin/{object_id}/{object_kind}/purge]
++ Parameters
+    + object_kind (string, required, `dds-file`) ... The kind of object.
+    + object_id (string, required, `777be35a-98e0-4c2e-9a17-7bc009f9b111`) ... The unique id of the object.
+
+### Purge Trashbin Item instance [PUT]
+>**Permission:** create_file or system_admin
+- - - - -
+**Response messages:**
+* 201: Valid API Token in 'Authorization' Header
```
Comment (dmann) [inline api_docs/trashbin.apib]: See above comment regarding `201`

File: api_docs/trashbin.apib [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+* 400: Parent object does not exist, has been purged, or is in the trash bin
+* 401: Missing, Expired, or Invalid API Token in 'Authorization' Header
+* 403: Forbidden
+* 404: Item does not exist in the trash bin
+
++ Response 201
+
```
Comment (dmann) [inline api_docs/trashbin.apib]: Wait... Where's the rest of the response? If this doesn't return a body, like a `DELETE` call, then it should probably return a `204`.

File: api_docs/trashbin.apib [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+>**Permission:** view_project or system_admin
+- - - - -
+
+>**Rules**
+- - - - -
++ Lists all children of the project unless the `name_contains` parameter is specified; when specified, only childen whose name contains the `name_contains` parameter are returned.
++ Lists only immediate children of the project unless the `recurse` parameter is specified, when specified, invokes a recursive search of the project hierarchy.
+
+**Response messages:**
+* 201: Valid API Token in 'Authorization' Header
```
Comment (dmann) [inline api_docs/trashbin.apib]: Should probably be `200: Success`

File: api_docs/trashbin.apib [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+                            "id": "ca29f7df-33ca-46dd-a015-92c46fdb6fd1",
+                            "name": "Knockout Mouse Project (KOMP)"
+                        },
+                        {
+                            "kind": "dds-folder",
+                            "id": "552be8c5-209d-4e3b-afaf-cb66686ffbyy",
+                            "name": "Sequencing Core"
+                        }
+                    ],
+                    "is_deleted": false,
```
Comment (dmann) [inline api_docs/trashbin.apib]: `s/false/true/`

File: api_docs/trashbin.apib [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
++ Response 200 (application/json)
+
+        {
+            "results": [
+                {
+                    "kind": "dds-folder",
+                    "id": "d5ae02a4-b9e6-473d-87c4-66f4c881ae7a",
+                    "parent": {
+                        "kind": "dds-folder",
```
Comment (dmann) [inline api_docs/trashbin.apib]: I'd say that the default for this endpoint is `recurse: false`. SO, these examples should have the project as the `parent`.

File: api_docs/trashbin.apib [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+                    "ancestors": [
+                        {
+                            "kind": "dds-project",
+                            "id": "ca29f7df-33ca-46dd-a015-92c46fdb6fd1",
+                            "name": "Knockout Mouse Project (KOMP)"
+                        },
+                        {
+                            "kind": "dds-folder",
+                            "id": "552be8c5-209d-4e3b-afaf-cb66686ffbyy",
+                            "name": "Sequencing Core"
```
Comment (dmann) [inline api_docs/trashbin.apib]: The `ancestors` should contain only the project.

File: api_docs/trashbin.apib [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+                            "id": "ca29f7df-33ca-46dd-a015-92c46fdb6fd1",
+                            "name": "Knockout Mouse Project (KOMP)"
+                        },
+                        {
+                            "kind": "dds-folder",
+                            "id": "2b91658a-2b0f-4ac6-83f5-287391610d0e",
+                            "name": "Sequencing Archive"
+                        }
+                    ],
+                    "is_deleted": false,
```
Comment (dmann) [inline api_docs/trashbin.apib]: `s/false/true/`

File: api_docs/trashbin.apib [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+                    "audit": { }
+                },
+                {
+                    "kind": "dds-file",
+                    "id": "777be35a-98e0-4c2e-9a17-7bc009f9b111",
+                    "parent": {
+                        "kind": "dds-folder",
+                        "id": "2b91658a-2b0f-4ac6-83f5-287391610d0e"
```
Comment (dmann) [inline api_docs/trashbin.apib]: See above comment regarding `parent`

File: api_docs/trashbin.apib [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+                    "ancestors": [
+                        {
+                            "kind": "dds-project",
+                            "id": "ca29f7df-33ca-46dd-a015-92c46fdb6fd1",
+                            "name": "Knockout Mouse Project (KOMP)"
+                        },
+                        {
+                            "kind": "dds-folder",
+                            "id": "2b91658a-2b0f-4ac6-83f5-287391610d0e",
+                            "name": "Sequencing Archive"
```
Comment (dmann) [inline api_docs/trashbin.apib]: See above comment regarding `ancestors`

File: api_docs/trashbin.apib [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+>**Permission:** view_project or system_admin
+- - - - -
+
+>**Rules**
+- - - - -
++ Lists all children of the folder unless the `name_contains` parameter is specified; when specified, only children whose name contains the `name_contains` parameter are returned.
++ Lists only immediate children of the folder unless the `recurse` parameter is specified, when specified, invokes a recursive search of the folder hierarchy.
+
+**Response messages:**
+* 201: Valid API Token in 'Authorization' Header
```
Comment (dmann) [inline api_docs/trashbin.apib]: Should probably be `200: Success`

File: api_docs/trashbin.apib [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+- - - - -
+
+>**Rules**
+- - - - -
++ Lists all children of the project unless the `name_contains` parameter is specified; when specified, only childen whose name contains the `name_contains` parameter are returned.
++ Lists only immediate children of the project unless the `recurse` parameter is specified, when specified, invokes a recursive search of the project hierarchy.
+
+**Response messages:**
+* 201: Valid API Token in 'Authorization' Header
+* 400: Parent object does not exist, has been purged, or is in the trash bin
```
Comment (dmann) [inline api_docs/trashbin.apib]: A `400` validation error can occur with a `GET`?

File: api_docs/trashbin.apib [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+- - - - -
+
+>**Rules**
+- - - - -
++ Lists all children of the folder unless the `name_contains` parameter is specified; when specified, only children whose name contains the `name_contains` parameter are returned.
++ Lists only immediate children of the folder unless the `recurse` parameter is specified, when specified, invokes a recursive search of the folder hierarchy.
+
+**Response messages:**
+* 201: Valid API Token in 'Authorization' Header
+* 400: Parent object does not exist, has been purged, or is in the trash bin
```
Comment (dmann) [inline api_docs/trashbin.apib]: A `400` validation error can occur with a `GET`?

File: api_docs/trashbin.apib [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+                            "id": "ca29f7df-33ca-46dd-a015-92c46fdb6fd1",
+                            "name": "Knockout Mouse Project (KOMP)"
+                        },
+                        {
+                            "kind": "dds-folder",
+                            "id": "552be8c5-209d-4e3b-afaf-cb66686ffbyy",
+                            "name": "Sequencing Core"
+                        }
+                    ],
+                    "is_deleted": false,
```
Comment (dmann) [inline api_docs/trashbin.apib]: `s/false/true/`

File: api_docs/trashbin.apib [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+                    "audit": { }
+                },
+                {
+                    "kind": "dds-file",
+                    "id": "777be35a-98e0-4c2e-9a17-7bc009f9b111",
+                    "parent": {
+                        "kind": "dds-folder",
+                        "id": "2b91658a-2b0f-4ac6-83f5-287391610d0e"
```
Comment (dmann) [inline api_docs/trashbin.apib]: Should probably have the same `parent` as the above folder.

File: api_docs/trashbin.apib [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+                    },
+                    "ancestors": [
+                        {
+                            "kind": "dds-project",
+                            "id": "ca29f7df-33ca-46dd-a015-92c46fdb6fd1",
+                            "name": "Knockout Mouse Project (KOMP)"
+                        },
+                        {
+                            "kind": "dds-folder",
+                            "id": "2b91658a-2b0f-4ac6-83f5-287391610d0e",
```
Comment (dmann) [inline api_docs/trashbin.apib]: Should probably have the same `parent` as the above folder.

File: api_docs/trashbin.apib [hunk 19]
```diff
@@ -0,0 +1,344 @@
...
+                            "name": "Sequencing Archive"
+                        }
+                    ],
+                    "is_deleted": false,
```
Comment (dmann) [inline api_docs/trashbin.apib]: `s/false/true/`

**Review by dmann — CHANGES_REQUESTED** [review]
I'm pretty sure that I got everything...

File: api_docs/trashbin.apib (hunk 18 above)
Comment (dmlond) [reply api_docs/trashbin.apib]: done

File: api_docs/trashbin.apib (hunk 2 above)
Comment (dmlond) [reply api_docs/trashbin.apib]: done

**Review by dmann — APPROVED** [review]

**Pull request merged**